File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/doxygen/DoxygenPreprocessor.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
package mil.tatrc.physiology.datamodel.doxygen;

import java.io.BufferedWriter;
import java.io.BufferedReader;
import java.nio.file.*;
import java.nio.charset.StandardCharsets;

import java.io.IOException;
import mil.tatrc.physiology.utilities.Log;

/**
 * 
 */
public class DoxygenPreprocessor
{
  public static void main(String[] args)
  {
    Log.setFileName("DoxygenPreprocessor.log");
    if(args.length!=2)
    { // Check # of input args
      Log.info("Command arguments are : <Directory of files to process> <Directory to place processed files>");
      return;
    }
    Path sourceDir = Paths.get(args[0]);
    if(!Files.isDirectory(sourceDir) ) { // locate source directory
      Log.error(String.format("Cannot find Source Directory : %s",args[0]));
      return;
    }
    Path destDir = Paths.get(args[1]);
    if(!Files.exists(destDir)) { // check for then create destination directory
      try {
        Files.createDirectories(destDir);
      } catch (IOException e) {
        Log.error (String.format("Unable to destination directory %s. I might not have file permissions", destDir)); 
      }
    }
    try {
      DirectoryStream<Path> directoryStream = Files.newDirectoryStream(sourceDir);
      for (Path file: directoryStream)  {        
        if( Files.isDirectory(file) || ! file.toString().toLowerCase().endsWith(".md") )
        {  continue; } // If the file is a directory or not an 'md' file, ignore it
        
        Log.info("Processing file : "+ file.toString() );
        BufferedReader inputSource = Files.newBufferedReader(file, StandardCharsets.ISO_8859_1);
        Path processedFile = destDir.resolve(file.getFileName());
        BufferedWriter processedResult      = Files.newBufferedWriter(processedFile,  StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
        String line = "";
        while ((line = inputSource.readLine()) != null) {
          if(line.indexOf("@insert")>-1)   {
            Path stubFile = Paths.get(line.substring(line.indexOf(" ")).trim());
            if( Files.exists(stubFile) ) {
              BufferedReader stubReader = Files.newBufferedReader(stubFile);            
              while ((line = stubReader.readLine()) != null) {  
                processedResult.write(line);   
                processedResult.newLine();
              }
              stubReader.close();
            } else if( Files.exists(Paths.get(args[0],stubFile.toString())) ) {
              // Read this file
              BufferedReader stubReader = Files.newBufferedReader(Paths.get(args[0]).resolve(stubFile));
              while ((line = stubReader.readLine()) != null)
              {  
                processedResult.write(line);  
                processedResult.newLine();
              }
              stubReader.close();
            } else {
              Log.error( String.format("Could not find %s to insert into this page searched [., %s] ", stubFile.toString(), args[0]));
              processedResult.newLine();
              processedResult.write("<img src=\"./images/MissingTable.jpg\"><center><i>Could not find "+stubFile.toString() +"</i></center><br>");
            }
          } else {
            processedResult.write(line);
            processedResult.newLine();
          }
        }
        inputSource.close();
        processedResult.close();
      }      
      directoryStream.close();
    } catch (IOException ex) {
        Log.error (ex);
        Log.error (String.format("An Exception has occured with the filesystem. I might not have file permissions"));

    }
  }
}

File: projects/python/cdm/src/mil/tatrc/physiology/datamodel/doxygen/DoxygenPreprocessor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Created on Sat Jun 27 12:20:10 2020

@author: Shashank

"""

import os
import argparse
import logging

class DoxygenPreprocessor():

    @classmethod
    def main(cls,args):
        if len(args)!=2:
            logging.error("Command arguments are : <Directory of files to process> <Directory to place processed files>")
            return
        else:
            if not os.path.isdir(args[0]):
                logging.error("Cannot find Source Directory : {0}".format(args[0]))
                return
            sourceDir=args[0]
            destDir=args[1]
            if not os.path.exists(args[1]):
                try:
                    os.mkdir(destDir)
                except IOError as e:
                    logging.warning("Unable to destination directory {0}. I might not have file permissions".format(destDir))
            try:
                for files in os.listdir(args[0]):
                    if os.path.isdir(files) or not files.endswith(".md"):
                        continue
                    print("Processing file : " + files)
                    fileprocess=os.path.join(args[0],files)
                    fileoutput=os.path.join(args[1],files)
                    with open(fileprocess,'r',encoding="iso-8859-1") as fin, open(fileoutput,'w',encoding="iso-8859-1") as fout:
                        for line in fin:
                            #print(line)
                            if line.startswith("@insert"):
                                path_file_1=line.partition("@insert")[2].strip(" ").strip("\n")
                                #print(path_file_1)
                                if not "/" in path_file_1:
                                    path_file_1=os.path.join(args[0],path_file_1)
                                if os.path.isfile(path_file_1):
                                    with open(path_file_1,'r',encoding="iso-8859-1") as fstub:
                                        for lines in fstub:
                                            fout.write(lines)
                                            #fout.write("\n")
                                    fstub.close()
                                else:
                                    print("Could not find {0} to insert into this page searched [., {0}] ".format(path_file_1))
                                    fout.write("\n")
                                    fout.write("<img src=\"./images/MissingTable.jpg\"><center><i>Could not find " + os.path.basename(path_file_1) + "</i></center><br>")
                            else:
                                fout.write(line)
                                #fout.write("\n")
                    fin.close()
                    fout.close()
            except IOError as ex:
                logging.warning(ex)
                logging.warning("An Exception has occured with the filesystem. I might not have file permissions")




if __name__=="__main__":
    import sys
    parser = argparse.ArgumentParser(description='Creation of all markdown files for generation of Documentation')
    parser.add_argument('-p','--markdownfile',help='Path to markdown or methodology files',required=True)
    parser.add_argument('-f','--Destdir',help='Path where files are needed to be pasted',required=True)
    args=parser.parse_args()
    if args.markdownfile == None or args.Destdir == None:
        parser.print_help()
        sys.exit(0)
    else:
        args=[args.markdownfile,args.Destdir]
        DoxygenPreprocessor.main(args)

